### PR TITLE
Improve Add OpCode performance

### PR DESCRIPTION
**Pull request by ahamlat**
## PR description
Improve Add opcode performance using new UInt256 implementation.

**Before this PR**
```
Benchmark                               Mode  Cnt    Score   Error  Units
AddOperationBenchmark.executeOperation  avgt   15  157.694 ± 2.838  ns/op
```

**With this PR**
```
Benchmark                               Mode  Cnt   Score   Error  Units
AddOperationBenchmark.executeOperation  avgt   15  84.921 ± 0.354  ns/op
```

This is almost **twice** faster 🚀 

## Fixed Issue(s)

https://github.com/hyperledger/besu/issues/9478

### Thanks for sending a pull request! Have you done the following?

- [x] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [x] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [x] spotless: `./gradlew spotlessApply`
- [x] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`
- [ ] hive tests: [Engine or other RPCs modified?](https://lf-hyperledger.atlassian.net/wiki/spaces/BESU/pages/22156302/Using+Hive+Test+Suite)

### CODE REVIEW

File: evm/src/main/java/org/hyperledger/besu/evm/MainnetEVMs.java [hunk 1]
```diff
@@ -219,12 +219,13 @@ private static void registerFrontierOperations(
       registry.put(new SModOperationOptimized(gasCalculator));
       registry.put(new AddModOperationOptimized(gasCalculator));
       registry.put(new MulModOperationOptimized(gasCalculator));
-
+      registry.put(new AddOperationOptimized(gasCalculator));
```
Comment (lu-pinto) [inline evm/src/main/java/org/hyperledger/besu/evm/MainnetEVMs.java]: you also need to add it to `EVM.java`

File: evm/src/test/java/org/hyperledger/besu/evm/UInt256Prop.java [hunk 2]
```diff
@@ -212,6 +212,98 @@ void property_divByZero_invariants() {
     assertThat(x.mulMod(x, zero).toBytesBE()).containsExactly(Bytes32.ZERO.toArrayUnsafe());
   }
 
+  // Simple ADD tests
+  // --------------------------------------------------------------------------
+  @Property
```
Comment (lu-pinto) [inline evm/src/test/java/org/hyperledger/besu/evm/UInt256Prop.java]: great tests! I like this framework

File: ethereum/core/src/jmh/java/org/hyperledger/besu/ethereum/vm/operations/AddOperationBenchmark.java [hunk 3]
```diff
@@ -15,13 +15,13 @@
 package org.hyperledger.besu.ethereum.vm.operations;
 
 import org.hyperledger.besu.evm.frame.MessageFrame;
-import org.hyperledger.besu.evm.operation.AddOperation;
+import org.hyperledger.besu.evm.operation.AddOperationOptimized;
 import org.hyperledger.besu.evm.operation.Operation;
 
 public class AddOperationBenchmark extends BinaryOperationBenchmark {
```
Comment (lu-pinto) [inline ethereum/core/src/jmh/java/org/hyperledger/besu/ethereum/vm/operations/AddOperationBenchmark.java:21]: nit: not sure if you want to add specific cases for addition instead of the default case of adding 2 fully randomized uint256 values.

File: evm/src/main/java/org/hyperledger/besu/evm/UInt256.java [hunk 4]
```diff
@@ -318,6 +318,17 @@ public UInt256 mulMod(final UInt256 other, final UInt256 modulus) {
     return new UInt256(result, modulus.length);
   }
 
+  /**
+   * Simple addition
+   *
+   * @param other The UInt256 to add to this.
+   * @return The UInt256 result from the addition
+   */
+  public UInt256 add(final UInt256 other) {
```
Comment (siladu) [inline evm/src/main/java/org/hyperledger/besu/evm/UInt256.java:450]: Thomas' original impl had these shortcuts, might be worth adding back in:
```
    if (other.isZero()) return this;
    if (this.isZero()) return other;
```

File: evm/src/test/java/org/hyperledger/besu/evm/UInt256Prop.java (hunk 2 above)
Comment (siladu) [reply evm/src/test/java/org/hyperledger/besu/evm/UInt256Prop.java]: Not sure if the prop tests explicitly cover it already, but is it worth adding an ua.add(ua) self-addition test?

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperationOptimized.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    final Bytes value0 = frame.popStackItem();
+    final Bytes value1 = frame.popStackItem();
+
+    UInt256 b0 = UInt256.fromBytesBE(value0.toArrayUnsafe());
+    UInt256 b1 = UInt256.fromBytesBE(value1.toArrayUnsafe());
+
+    UInt256 result = b0.add(b1);
+    byte[] resultArray = result.toBytesBE();
+    frame.pushStackItem(Bytes.wrap(resultArray));
```
Comment (siladu) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperationOptimized.java:61]: Thomas' original impl had overflow wrapping. Not sure if this is relevant with latest iteration. I think toBytesBE() effectively limits to 8 limbs, but wonder if we should be defensive here as well?

```
    int length = resultArray.length;
    if (length > 32) {
      frame.pushStackItem(Bytes.wrap(resultArray, length - 32, 32));
    } else {
      frame.pushStackItem(Bytes.wrap(resultArray));
    }
```

Comment (ahamlat) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperationOptimized.java:61]: IMO, this is handled by the inner design in UInt256. The new implementation removed those checks in current MOD optimized version.

File: evm/src/main/java/org/hyperledger/besu/evm/UInt256.java (hunk 4 above)
Comment (ahamlat) [reply evm/src/main/java/org/hyperledger/besu/evm/UInt256.java:450]: For these opCodes (<5 gas), we should keep the implementation very simple and avoid more checks. In this suggestion, to optimize one specifc "uncommon" case, we have to penalize all other cases.

File: ethereum/core/src/jmh/java/org/hyperledger/besu/ethereum/vm/operations/AddOperationBenchmark.java (hunk 3 above)
Comment (ahamlat) [reply ethereum/core/src/jmh/java/org/hyperledger/besu/ethereum/vm/operations/AddOperationBenchmark.java:21]: I don't think it is needed TBH, because the algorithm behind is very simple. I understand to have more test cases for an algorithm like MOD with some branches and try to hit different execution. On the add opcode I don't think it is needed.

File: evm/src/test/java/org/hyperledger/besu/evm/UInt256Prop.java (hunk 2 above)
Comment (ahamlat) [reply evm/src/test/java/org/hyperledger/besu/evm/UInt256Prop.java]: Yeah, I can test that and check that it returns twice ua.

File: evm/src/main/java/org/hyperledger/besu/evm/MainnetEVMs.java (hunk 1 above)
Comment (ahamlat) [reply evm/src/main/java/org/hyperledger/besu/evm/MainnetEVMs.java]: Good catch.

File: evm/src/test/java/org/hyperledger/besu/evm/UInt256Prop.java (hunk 2 above)
Comment (ahamlat) [reply evm/src/test/java/org/hyperledger/besu/evm/UInt256Prop.java]: https://github.com/hyperledger/besu/pull/9477/commits/1577b854c8b020a9826c6ebfa4582a992eff31f8

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperationOptimized.java (hunk 5 above)
Comment (lu-pinto) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperationOptimized.java:61]: agree it should be handled by the UInt256 implementation if it's not handled it should be fixed there.
From the spec I understand that bytes should just be truncated. Regarding padding you don't need since you always have 8 fixed limbs (including zeros).

Comment (siladu) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperationOptimized.java:61]: I think it's handled when you use default 8 limbs, though maybe not as explicitly. Not sure what would happen if you choose to use > 8 limbs.

Comment (ahamlat) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperationOptimized.java:61]: I added more property based testing on the boundaries to be sure that this implementation is working as expected :  https://github.com/hyperledger/besu/pull/9477/commits/278fac9255e8c61d6d167276ad28ce704ca55479

**Comment by ahamlat** [comment]
Please don't merge until a hoodi full sync is finished.

**Comment by ahamlat** [comment]
So Hoodi full sync is not stable with this PR, I need to investigage. I'm going to make draft until I have a clear understanding on what's going on Hoodi.

**Comment by ahamlat** [comment]
After investigation, instability on hoodi full sync was related to a regression in Hoodi full sync performance. That regression was reverted in this PR https://github.com/hyperledger/besu/pull/9580.
